Let `supergraph compose --output` support paths with missing intermediate directories

jonathanrainer opened the pull request:
Fixes #1787 

At the moment if a path is presented to `supergraph compose --output` where an intermediate directory in the output doesn't exist. It just fails silently, which isn't good. This allows those commands to succeed, creating the intermediate directories in the process.

jonathanrainer committed:
Check if a parent path exists and create it if not

Also update how we open and write to files, so we can do less manual
checking

jonathanrainer committed:
Fix messaging

jonathanrainer committed:
Fix up clippies

aaronArinder reviewed (approved):
I think this looks legit; how'd you test it, trying it out locally?

jonathanrainer committed:
Second attempt with added unit testing

File system operations are so complicated and nuanced we need actual
testing to prove what we're doing is sensible. Hence the rejig
here.

jonathanrainer committed:
Sure up commentary

dotdat commented on crates/rover-std/src/fs.rs:
```diff
@@ -54,23 +55,70 @@ impl Fs {
+            return Err(anyhow::anyhow!("cannot write file to root or prefix {}", path).into());
+        }?;
+
+        return if let Some(file_name) = path.file_name() {
```
nit: seems like this would be more idiomatic as:
```
match path.file_name() {
  Some(filename) => {
    ...
    Ok(())
  }
  None => { 
    ...
    Err(...)
  }
}
```
Putting this in the return position (without the semicolon) and using a `match` statement seems closer to home.

aaronArinder commented on crates/rover-std/src/fs.rs:
```diff
@@ -54,23 +55,70 @@ impl Fs {
+            Err(anyhow::anyhow!("cannot write to a path without a final element {}", path).into())
+        };
+    }
+
```
I think there might be a way to write this in a more top-down readable-y way (at least, I got stuck on how the `return` was placed with the nested `OK(())` and `Err(..)`, but this might just by my idiosyncrasies)

here's a quick switcheroo for how to make it read more linerally:
```
    {
        let path = path.as_ref();
        tracing::info!("checking existence of parent path in '{}'", path);
        // Grab the parent path then attempt to canonicalize it, we can't just canonicalize the
        // entire path because that would entail the file existing, which of course it doesn't yet.
        //
        // If we can't pull out a final element, then we're in a weird case but should bail
        // out as there's nothing sensible we can do.
        let file_name = path.file_name().ok_or(anyhow::anyhow!(
            "cannot write to a path without a final element {path}"
        ))?;

        let mut canonical_final_path = path
            .parent()
            .map(Self::upsert_path_exists)
            // If the has no parent, the only conceivable case is that we are looking at the root,
            // and as that cannot be a file we should bail out at this point
            .ok_or(anyhow::anyhow!(
                "cannot write file to root or prefix {path}"
            ))??;

        canonical_final_path.push(file_name);

        tracing::debug!("final canonical path is {}", canonical_final_path);

        let mut file = OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)
            .with_context(|| {
                format!(
                    "tried to open {} but was unable to do so",
                    &canonical_final_path
                )
            })?;

        tracing::info!("writing {} to disk", &canonical_final_path);

        // Actually write the file out to where it needs to be
        file.write(contents.as_ref())
            .with_context(|| format!("could not write {}", &canonical_final_path))?;

        Ok(())
    }
```

I moved the `debug` log for the attempt to canonicalize into the `upsert_path_exists` fn

jonathanrainer replied:
This is a great shout, thanks!

aaronArinder commented on crates/rover-std/src/fs.rs:
```diff
@@ -54,23 +55,70 @@ impl Fs {
+
+    /// Given a path, where some elements may not exist, it will return the canonical
+    /// representation of the path, AND create any missing interim directories.
+    fn upsert_path_exists(path: &Utf8Path) -> Result<Utf8PathBuf, anyhow::Error> {
```
does this need to be a method off of `Fs`? trying to figure out if it adds anything semantically to have it as part of `Fs` or if folks will get confused seeing it there

aaronArinder replied:
I guess there's a precedent with `create_dir_all`, but I sort of think _that_ fn is confusing to have under `Fs` too, but I guess it gets used throughout the codebase; makes me glad that `upsert_path_exists` doesn't need to be marked `pub`

jonathanrainer replied:
I feel like since it's a private method that's kind of ok? You could in-line it, but it felt like the method it was part of was complex enough without adding a big closure in the middle. Happy to change it though if you think that's better?

aaronArinder replied:
I really don't know! I don't think it _really_ matters either way; it's also worth calling out that I don't know much about the conventions rover has been using in terms of code organization

aaronArinder reviewed (approved):
re-reviewed and still looks good; by my lights, the snippet I added is a hair more readable, but it's not a deal-breaker for my review if you don't think it's more readable

jonathanrainer committed:
Re-write main function so its more readable

aaronArinder reviewed (approved):
lgtm!
